### PR TITLE
Conditionally-logging interceptor sample

### LINKED ISSUE #4156: Set logging interceptor based on API
As projects grow the number of API requests sent by the application also grows, and hence the debugging of APIs gets more complicated. It would be much convenient to be able to specify API calls by some annotation to output okhttp log. Something like this:

```
@LogInterceptor(DefaultHttpLogInterceptor())
@POST("/users/sign-in-phone")
    suspend fun loginPhone(@Body body: LoginPhoneRequestDto): Response<LoginResponseDTO.LoginSuccessfulResponseDTO>
```
This would print out the log of this specific API instead of whole network connection to server. 
Is it feasible to add such thing to Retrofit?

### PR DESCRIPTION
**Pull request by JakeWharton**
Closes #4156 

---

- [ ] `CHANGELOG.md`'s "Unreleased" section has been updated, if applicable.

### CODE REVIEW

**Comment by JakeWharton** [comment]
Output:
```
May 29, 2024 11:40:53 AM okhttp3.mockwebserver.MockWebServer$3 execute
INFO: MockWebServer[63842] starting to accept connections
May 29, 2024 11:40:53 AM okhttp3.mockwebserver.MockWebServer$4 processOneRequest
INFO: MockWebServer[63842] received request: GET /one HTTP/1.1 and responded: HTTP/1.1 200 OK
May 29, 2024 11:40:53 AM okhttp3.mockwebserver.MockWebServer$4 processOneRequest
INFO: MockWebServer[63842] received request: GET /two HTTP/1.1 and responded: HTTP/1.1 200 OK
--> GET http://localhost:63842/two
--> END GET
<-- 200 OK http://localhost:63842/two (0ms)
Content-Length: 0
<-- END HTTP (0-byte body)
```

File: samples/src/main/java/com/example/retrofit/ConditionalLoggingInterceptor.kt [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  @LogInterceptor
+  @GET("two")
+  suspend fun two(): ResponseBody
+}
+
+/**
+ * Retrofit service functions which are annotated with this class will have their HTTP calls
+ * logged. You must add [ConditionalLoggingInterceptor] to your [OkHttpClient] for this to work.
+ */
+annotation class LogInterceptor
```
Comment (mattprecious) [inline samples/src/main/java/com/example/retrofit/ConditionalLoggingInterceptor.kt]: nit: Should this just be `Log`? It's kind of weird to think about interceptors as a concept in your service definition.

File: samples/src/main/java/com/example/retrofit/ConditionalLoggingInterceptor.kt [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+class ConditionalLoggingInterceptor(
+  private val loggingInterceptor: HttpLoggingInterceptor,
+) : Interceptor {
+  override fun intercept(chain: Interceptor.Chain): Response {
+    val request = chain.request()
+    request.tag(Invocation::class.java)?.let { invocation ->
+      if (invocation.method().isAnnotationPresent(LogInterceptor::class.java)) {
```
Comment (mattprecious) [inline samples/src/main/java/com/example/retrofit/ConditionalLoggingInterceptor.kt]: Does the annotation need to have a specific retention for this to reliably work? Also, only an Android issue, but would r8 rules be needed as well?

**Review by mattprecious — APPROVED** [review]

File: samples/src/main/java/com/example/retrofit/ConditionalLoggingInterceptor.kt (hunk 2 above)
Comment (JakeWharton) [reply samples/src/main/java/com/example/retrofit/ConditionalLoggingInterceptor.kt]: The default is runtime retention which should be fine. These are only samples so R8 is an exercise left for the reader. It should generally be okay without any rules, I think.

File: samples/src/main/java/com/example/retrofit/ConditionalLoggingInterceptor.kt (hunk 1 above)
Comment (JakeWharton) [reply samples/src/main/java/com/example/retrofit/ConditionalLoggingInterceptor.kt]: I started with `Log` and it looked weird, but I don't care. Put back.

**Pull request merged**